src/main/java/hudson/plugins/jacoco/JacocoPublisher.java

```diff
@@ -259,142 +382,121 @@ protected static String resolveFilePaths(AbstractBuild<?, ?> build, BuildListene
         return input;
     }
 	
-	protected static FilePath[] resolveDirPaths(AbstractBuild<?, ?> build, BuildListener listener, final String input) {
+    protected static FilePath[] resolveDirPaths(FilePath workspace, TaskListener listener, final String input) {
 		//final PrintStream logger = listener.getLogger();
 		FilePath[] directoryPaths = null;
 		try {
-			directoryPaths = build.getWorkspace().act(new FilePath.FileCallable<FilePath[]>() 
-			{
-				static final long serialVersionUID = 1552178457453558870L;
-
-				public FilePath[] invoke(File f, VirtualChannel channel) throws IOException {
-					FilePath base = new FilePath(f);
-					ArrayList<FilePath> localDirectoryPaths= new ArrayList<FilePath>();
-					String[] includes = input.split(DIR_SEP);
-					DirectoryScanner ds = new DirectoryScanner();
-			        
-			        ds.setIncludes(includes);
-			        ds.setCaseSensitive(false);
-			        ds.setBasedir(f);
-			        ds.scan();
-			        String[] dirs = ds.getIncludedDirectories();
-			        
-			        for (String dir : dirs) {
-                        localDirectoryPaths.add(base.child(dir));
-			        }
-			        FilePath[] lfp = {};//trick to have an empty array as a parameter, so the returned array will contain the elements
-			        return localDirectoryPaths.toArray(lfp);
-	            }
-			});
-			
+            directoryPaths = workspace.act(new ResolveDirPaths(input));
 		} catch(InterruptedException ie) {
 			ie.printStackTrace();
 		} catch(IOException io) {
 			io.printStackTrace();
 		}
 		return directoryPaths;
 	}
-	
-	/* 
-	 * Entry point of this report plugin.
-	 * 
-	 * @see hudson.tasks.BuildStepCompatibilityLayer#perform(hudson.model.AbstractBuild, hudson.Launcher, hudson.model.BuildListener)
-	 */
-    @SuppressWarnings("resource")
+
+
     @Override
-    public boolean perform(AbstractBuild<?,?> build, Launcher launcher, BuildListener listener) throws InterruptedException, IOException {
-	
-		final PrintStream logger = listener.getLogger();
-		FilePath[] matchedClassDirs = null;
-		FilePath[] matchedSrcDirs = null;
+    public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath filePath, @Nonnull Launcher launcher, @Nonnull TaskListener taskListener) throws InterruptedException, IOException {
+        Map<String, String> envs = run instanceof AbstractBuild ? ((AbstractBuild) run).getBuildVariables() : Collections.<String, String>emptyMap();
+
+        healthReports = createJacocoHealthReportThresholds();
+
+        final PrintStream logger = taskListener.getLogger();
+        FilePath[] matchedClassDirs = null;
+        FilePath[] matchedSrcDirs = null;
+
+        if (run.getResult() == Result.FAILURE || run.getResult() == Result.ABORTED) {
+            return;
+        }
+
+
+        logger.println("[JaCoCo plugin] Collecting JaCoCo coverage data...");
+
+
+        EnvVars env = run.getEnvironment(taskListener);
```

Comment:
Note that due to [JENKINS-29144](https://issues.jenkins-ci.org/browse/JENKINS-29144) this will not handle Pipeline `withEnv`.

Anyway `env` seems to be unused. I think you meant to pass it to `resolveFilePaths`.

Anyway, for Pipeline, we do not _want_ anything to be substituted. `execPattern` should be taken as a literal. So what you probably actually wanted was something like

```java
String _execPattern;
if (build instanceof AbstractBuild) {
    _execPattern = resolveFilePaths((AbstractBuild) build, listener, execPattern);
} else {
    _execPattern = execPattern;
}
…use as: filePath.list(_execPattern)
```

where you fix the helper method

```java
protected static String resolveFilePaths(AbstractBuild<?, ?> build, TaskListener listener, String input) {
    EnvVars env = build.getEnvironment(listener);
    env.overrideAll(build.getBuildVariables());
    // …as before
```